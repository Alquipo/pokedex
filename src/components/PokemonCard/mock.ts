export default [
  {
    id: 6,
    name: 'Charizard',
    types: ['Fire', 'Flying']
  },
  {
    id: 7,
    name: 'Squirtle',
    types: ['Water']
  },
  {
    id: 8,
    name: 'Wartortle',
    types: ['Water']
  },
  {
    id: 9,
    name: 'Blastoise',
    types: ['Water']
  },
  {
    id: 10,
    name: 'Cartepie',
    types: ['Bug']
  },
  {
    id: 11,
    name: 'Metapod',
    types: ['Bug']
  },
  {
    id: 12,
    name: 'Butterfree',
    types: ['Bug', 'Flying']
  },
  {
    id: 13,
    name: 'Weedle',
    types: ['Bug', 'Poison']
  },
  {
    id: 14,
    name: 'Kakuna',
    types: ['Bug', 'Poison']
  },
  {
    id: 15,
    name: 'Beedrill',
    types: ['Bug', 'Poison']
  },
  {
    id: 16,
    name: 'Pidgey',
    types: ['Normal', 'Flying']
  },
  {
    id: 6,
    name: 'Charizard',
    types: ['Fire', 'Flying']
  },
  {
    id: 7,
    name: 'Squirtle',
    types: ['Water']
  },
  {
    id: 8,
    name: 'Wartortle',
    types: ['Water']
  },
  {
    id: 9,
    name: 'Blastoise',
    types: ['Water']
  },
  {
    id: 10,
    name: 'Cartepie',
    types: ['Bug']
  },
  {
    id: 11,
    name: 'Metapod',
    types: ['Bug']
  },
  {
    id: 12,
    name: 'Butterfree',
    types: ['Bug', 'Flying']
  },
  {
    id: 13,
    name: 'Weedle',
    types: ['Bug', 'Poison']
  },
  {
    id: 14,
    name: 'Kakuna',
    types: ['Bug', 'Poison']
  },
  {
    id: 15,
    name: 'Beedrill',
    types: ['Bug', 'Poison']
  }
]
