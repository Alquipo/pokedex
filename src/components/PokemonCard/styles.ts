import styled, { css, DefaultTheme } from 'styled-components'
import { PokemonTypesProps } from '.'
import media from 'styled-media-query'

export type TypeColorProps = {
  typeColor: PokemonTypesProps
}

export const wrapperModifiers = {
  type: (theme: DefaultTheme, typeColor: PokemonTypesProps) => css`
    border-color: ${theme.colors.pokemonType[typeColor]};
    color: ${theme.colors.pokemonType[typeColor]};

    &:hover {
      box-shadow: 0 0 1rem ${theme.colors.pokemonType[typeColor]};
      color: ${theme.colors.pokemonType[typeColor]};
    }
  `
}

export const Wrapper = styled.div<TypeColorProps>`
  ${({ theme, typeColor }) => css`
    width: auto;
    position: relative;

    padding: ${theme.spacings.xxsmall};
    margin: ${theme.spacings.xxsmall} ${theme.spacings.xsmall};

    border-radius: ${theme.border.radius.large};
    border: 0.1rem solid;
    text-align: center;

    cursor: pointer;

    background-color: ${theme.colors.white};
    text-decoration: none;

    transition: all 0.2s ease-in-out;

    ${media.greaterThan('medium')`
      &:hover {
          transform: scale(1.01);
        }
    `}

    ${!!typeColor && wrapperModifiers.type(theme, typeColor)}
  `}
`
export const CardName = styled.h1`
  ${({ theme }) => css`
    text-transform: capitalize;
    font-size: ${theme.font.sizes.xlarge};
    font-weight: ${theme.font.bold};
    width: 19rem;
  `}
`

export const WrapperIcons = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
`
