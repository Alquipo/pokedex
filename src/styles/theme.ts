export default {
  grid: {
    container: '140rem',
    gutter: '3.2rem'
  },

  border: {
    radius: {
      xsmall: '0.3rem',
      small: '0.5rem',
      large: '1.0rem'
    }
  },

  font: {
    family:
      "Poppins, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    medium: 600,
    normal: 400,
    bold: 700,
    sizes: {
      xsmall: '1.2rem',
      small: '1.4rem',
      medium: '1.6rem',
      large: '1.8rem',
      xlarge: '2.0rem',
      xxlarge: '3.5rem'
    }
  },

  colors: {
    // primary: '#ecf0f1',
    // secondary: '#e4c439',
    // mainBg: '#06092B',
    // black: '#030517',
    // darkGray: '#2E2F42',
    lightBg: '#f2f3f5',
    white: '#fff',
    lightGray: '#EAEAEA',
    gray: '#505050',

    pokemonType: {
      normal: '#A0A29F',
      fire: '#FBA54C',
      water: '#539DDF',
      electric: '#F2D94E',
      grass: '#5FBD58',
      ice: '#75D0C1',
      ground: '#DA7C4D',
      flying: '#A1BBEC',
      ghost: '#5F6DBC',
      rock: '#C9BB8A',
      fighting: '#D3425F',
      poison: '#B763CF',
      psychic: '#FA8581',
      bug: '#92BC2C',
      dark: '#595761',
      steel: '#5695A3',
      dragon: '#0C69C8',
      fairy: '#EE90E6'
    },

    pokemonText: {
      normal: '#bcaaa4',
      fire: '#ff8a80',
      water: '#81d4fa',
      electric: '#ffd600',
      grass: '#81c784',
      poison: '#ad8ee7',
      fairy: '#f8bbd0',

      bug: '#81c784',
      ice: '#81d4fa',
      ground: '#bcaaa4',
      flying: '#81d4fa',
      ghost: '#ad8ee7',
      rock: '#bcaaa4',
      fighting: '#ff8a80',
      psychic: '#ff8a80',
      dark: '#bcaaa4',
      steel: '#5695A3',
      dragon: '#81d4fa',

      transparent: 'hsla(0,0%,100%,0.5)'
    }
  },

  transition: {
    color: 'color 1000ms ease-in, width 1500ms ease-in-out',
    backgroundColor: 'background-color 1000ms ease-in, width 1500ms ease-in-out'
  },

  spacings: {
    xxsmall: '0.3rem',
    xsmall: '0.8rem',
    small: '1.6rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '7.5rem'
  },

  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  }
} as const
