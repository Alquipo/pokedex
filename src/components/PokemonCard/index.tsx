import Image from 'next/image'

import PokemonTypeIcon from 'components/PokemonTypeIcon'
import Ribbon, { RibbonPosition } from 'components/Ribbon'
import { formatNameID, formatNameImageCompressed } from 'utils/formatName'
import * as S from './styles'

export type PokemonTypesProps =
  | 'normal'
  | 'fire'
  | 'water'
  | 'electric'
  | 'grass'
  | 'ice'
  | 'ground'
  | 'flying'
  | 'ghost'
  | 'rock'
  | 'fighting'
  | 'poison'
  | 'psychic'
  | 'bug'
  | 'dark'
  | 'steel'
  | 'dragon'
  | 'fairy'

export type PokemonCardProps = {
  id: number
  name: string
  types: PokemonTypesProps[]
  ribbonPosition?: RibbonPosition
}

const PokemonCard = ({ id, name, types, ribbonPosition }: PokemonCardProps) => (
  <S.Wrapper typeColor={types[0]}>
    <Ribbon position={ribbonPosition} color={types[0]}>
      #{formatNameID(id)}
    </Ribbon>

    <Image
      src={formatNameImageCompressed(id)}
      alt={name}
      width={100}
      height={100}
      role="img"
    />
    <S.CardName>{name}</S.CardName>

    <S.WrapperIcons>
      {types.map((type, index) => (
        <PokemonTypeIcon hasHover key={index} type={type} />
      ))}
    </S.WrapperIcons>
  </S.Wrapper>
)

export default PokemonCard
