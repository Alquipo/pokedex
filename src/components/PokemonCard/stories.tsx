import { Story, Meta } from '@storybook/react/types-6-0'
import PokemonCard, { PokemonCardProps, PokemonTypesProps } from '.'

export type RibbonColors = PokemonTypesProps

export default {
  title: 'Home/PokemonCard',
  component: PokemonCard,
  args: {
    id: 6,
    name: 'pikachu',
    types: ['fire', 'flying']
  },
  argTypes: {
    types: {
      control: {
        type: 'inline-check',
        options: [
          'normal',
          'fire',
          'water',
          'electric',
          'grass',
          'ice',
          'ground',
          'flying',
          'ghost',
          'rock',
          'fighting',
          'poison',
          'psychic',
          'bug',
          'dark',
          'steel',
          'dragon',
          'fairy'
        ]
      }
    }
  }
} as Meta

export const Default: Story<PokemonCardProps> = (args) => (
  <div
    style={{
      display: 'flex',
      justifyContent: 'center'
    }}
  >
    <PokemonCard {...args} />
  </div>
)
