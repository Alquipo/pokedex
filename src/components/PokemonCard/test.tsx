import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import theme from 'styles/theme'

import PokemonCard, { PokemonCardProps } from '.'

const props: PokemonCardProps = {
  id: 6,
  name: 'Charizard',
  types: ['fire', 'flying']
}

jest.mock('next/image', () => ({
  __esModule: true,
  default: function Mock({ children }: { children: React.ReactNode }) {
    return <img data-testid="Mock Image">{children}</img>
  }
}))

describe('<PokemonCard />', () => {
  it('should render the PokemonCard correctly', () => {
    const { container } = renderWithTheme(<PokemonCard {...props} />)

    expect(screen.getByText('Charizard')).toBeInTheDocument()

    expect(screen.getByText('#006')).toBeInTheDocument()

    expect(screen.getAllByTestId('Mock Image')).toHaveLength(3)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a Ribbon', () => {
    renderWithTheme(<PokemonCard {...props} />)

    const ribbon = screen.getByText(/#006/i)

    expect(ribbon).toBeInTheDocument()
    expect(ribbon).toHaveStyle({
      backgroundColor: theme.colors.pokemonType.fire
    })
  })
})
