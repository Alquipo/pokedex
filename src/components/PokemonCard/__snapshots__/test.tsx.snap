// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonCard /> should render the PokemonCard correctly 1`] = `
.c4 {
  border-radius: 100%;
  height: 3rem;
  width: 3rem;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  margin: 2px 3px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FBA54C;
  box-shadow: 0 0 2rem #FBA54C;
}

.c5 {
  border-radius: 100%;
  height: 3rem;
  width: 3rem;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  margin: 2px 3px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #A1BBEC;
  box-shadow: 0 0 2rem #A1BBEC;
}

.c1 {
  position: absolute;
  padding: 0 0.8rem;
  height: 2.6rem;
  top: 0.3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.2rem;
  font-weight: 700;
  color: #fff;
  background-color: #FBA54C;
  left: -1.5rem;
}

.c1::before {
  content: '';
  top: 2.6rem;
  border-top-width: 0.7rem;
  position: absolute;
  border-style: solid;
  border-bottom-color: transparent;
  border-bottom-width: 1rem;
}

.c1::before {
  border-left-color: #dc7205;
  border-top-color: #dc7205;
}

.c1::before {
  border-left-width: 1.5rem;
  border-right-width: 0rem;
  border-left-color: transparent;
  left: 0;
}

.c0 {
  width: auto;
  position: relative;
  padding: 0.3rem;
  margin: 0.3rem 0.8rem;
  border-radius: 1.0rem;
  border: 0.1rem solid;
  text-align: center;
  cursor: pointer;
  background-color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  border-color: #FBA54C;
  color: #FBA54C;
}

.c0:hover {
  box-shadow: 0 0 1rem #FBA54C;
  color: #FBA54C;
}

.c2 {
  text-transform: capitalize;
  font-size: 2.0rem;
  font-weight: 700;
  width: 19rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media (min-width:768px) {
  .c4:hover {
    -webkit-filter: saturate(200%);
    filter: saturate(200%);
    -webkit-transform: scale(1.1);
    -ms-transform: scale(1.1);
    transform: scale(1.1);
  }
}

@media (min-width:768px) {
  .c5:hover {
    -webkit-filter: saturate(200%);
    filter: saturate(200%);
    -webkit-transform: scale(1.1);
    -ms-transform: scale(1.1);
    transform: scale(1.1);
  }
}

@media (min-width:768px) {
  .c0:hover {
    -webkit-transform: scale(1.01);
    -ms-transform: scale(1.01);
    transform: scale(1.01);
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    color="fire"
  >
    #
    006
  </div>
  <img
    data-testid="Mock Image"
  />
  <h1
    class="c2"
  >
    Charizard
  </h1>
  <div
    class="c3"
  >
    <div
      class="c4"
    >
      <img
        data-testid="Mock Image"
      />
    </div>
    <div
      class="c5"
    >
      <img
        data-testid="Mock Image"
      />
    </div>
  </div>
</div>
`;
